scheduler/src/tbf/ratelimit.clj

```diff
@@ -0,0 +1,101 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns tbf.ratelimit
+  (:require
+    [tbf.core :as tbf-ns])
+  (:import (com.google.common.cache CacheBuilder Cache)
+           (java.util.concurrent TimeUnit)))
+
+(defprotocol RateLimiter
+  (earn-tokens [this key]
+    "Account for any earned tokens for the requested key.")
+  (in-debt [this key]
+    "Boolean indicating if the rate limit is currently in debt. Note that this is read-only and does
+    not earn tokens. For an up-to-date value, you will want to earn tokens first.")
+  (take-time-until [this key tokens]
+    "Request the number of tokens. Returns the number of milliseconds one should wait until those tokens should be spent.
+    (I.e. if currently in debt at the time of the request, before the request is actually processed, this will tell
+    you how long until you get out of debt.)")
+  (time-until [this key]
+    "Make a rate limiter module. Each bucket has a maximum capacity, earns at the requested rate per minute.
+     Old unused buckets expire after the given number of minutes."))
+
+(defn get-tbf
+      "Get or create a TBF for the requested key. Buckets start off half-full, so users whose first request is a lot of
+      fast requests don't get throttled immediately. The fact that we don't have a TBF says that we are either just starting
+      up (not likely) or haven't gotten anything from them in hours."
+      [limiter key]
+      (.get (:cache limiter) key (proxy [Callable] []
+                                   (call []
+                                     (tbf-ns/make (/ (:rate-per-minute limiter) 60000.)
+                                                  (:max-tokens limiter)
+                                                  (System/currentTimeMillis)
+                                                  (/ (:max-tokens limiter) 2))))))
+
+(defrecord TBF-RateLimiter
+  [^Long max-tokens ^Float rate-per-minute cache]
+  RateLimiter
+  (earn-tokens
+    [this key]
+    (locking cache)
```

Comment:
The `let` block needs to be inside the `locking` block.

Reply:
Good catch!